servers/src/main/java/tachyon/master/block/BlockMaster.java

```diff
@@ -0,0 +1,571 @@
+import tachyon.thrift.BlockLocation;
+import tachyon.thrift.BlockMasterService;
+import tachyon.thrift.Command;
+import tachyon.thrift.CommandType;
+import tachyon.thrift.NetAddress;
+import tachyon.thrift.WorkerInfo;
+import tachyon.util.CommonUtils;
+import tachyon.util.FormatUtils;
+import tachyon.util.ThreadFactoryUtils;
+import tachyon.util.io.PathUtils;
+
+// TODO (Gene): Add javadoc
+public final class BlockMaster extends MasterBase implements ContainerIdGenerator {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  // Block metadata management.
+  /**
+   * blocks ever placed in all Tachyon workers, including both the active blocks and lost ones. This
+   * state much be journaled.
+   */
+  private final Map<Long, MasterBlockInfo> mBlocks = new HashMap<Long, MasterBlockInfo>();
+  /**
+   * This state much be journaled.
+   */
+  private final BlockIdGenerator mBlockIdGenerator = new BlockIdGenerator();
+  private final Set<Long> mLostBlocks = new HashSet<Long>();
+  private final BlockingQueue<MasterWorkerInfo> mLostWorkers =
+      new LinkedBlockingQueue<MasterWorkerInfo>();
+  private final TachyonConf mTachyonConf;
+
+  /**
+   * the service that detects lost worker nodes, and tries to restart the failed workers
+   */
+  private Future<?> mLostWorkerDetectionService;
+
+  // Worker metadata management.
+  private final IndexedSet.FieldIndex<MasterWorkerInfo> mIdIndex =
+      new IndexedSet.FieldIndex<MasterWorkerInfo>() {
+        @Override
+        public Object getFieldValue(MasterWorkerInfo o) {
+          return o.getId();
+        }
+      };
+  private final IndexedSet.FieldIndex<MasterWorkerInfo> mAddressIndex =
+      new IndexedSet.FieldIndex<MasterWorkerInfo>() {
+        @Override
+        public Object getFieldValue(MasterWorkerInfo o) {
+          return o.getAddress();
+        }
+      };
+
+  @SuppressWarnings("unchecked")
+  private final IndexedSet<MasterWorkerInfo> mWorkers =
+      new IndexedSet<MasterWorkerInfo>(mIdIndex, mAddressIndex);
+  // This state must be journaled.
+  private final AtomicLong mNextWorkerId = new AtomicLong(1);
+
+  public static String getJournalDirectory(String baseDirectory) {
+    return PathUtils.concatPath(baseDirectory, Constants.BLOCK_MASTER_SERVICE_NAME);
+  }
+
+  public BlockMaster(Journal journal, TachyonConf tachyonConf) {
+    super(journal,
+        Executors.newFixedThreadPool(2, ThreadFactoryUtils.build("block-master-%d", true)));
+    mTachyonConf = tachyonConf;
+  }
+
+  @Override
+  public TProcessor getProcessor() {
+    return new BlockMasterService.Processor<BlockMasterServiceHandler>(
+        new BlockMasterServiceHandler(this));
+  }
+
+  @Override
+  public String getProcessorName() {
+    return Constants.BLOCK_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  public void processJournalCheckpoint(JournalInputStream inputStream) throws IOException {
+    // clear state before processing checkpoint.
```

Comment:
it is not necessarily corrupted. If it were a standby master, it would have some state, and then if it had to process a different checkpoint file, it would have to clear the state.